Code/GraphMol/SynthonSpaceSearch/SynthonSpaceFingerprintSearcher.cpp

```diff
@@ -0,0 +1,151 @@
+      if (const auto sim = TanimotoSimilarity(*fragFPs[i], *synthonFPs[j]);
+          sim >= similarityCutoff) {
+        synthonsToUse[synthonOrder[i]][j] = true;
+        fragMatched = true;
+      }
+    }
+    if (!fragMatched) {
+      synthonsToUse.clear();
+      return synthonsToUse;
+    }
+  }
+
+  // Fill in any synthons where they all didn't match.
+  details::expandBitSet(synthonsToUse);
+  return synthonsToUse;
+}
+}  // namespace
+
+std::vector<SynthonSpaceHitSet> SynthonSpaceFingerprintSearcher::searchFragSet(
+    std::vector<std::unique_ptr<ROMol>> &fragSet) const {
+  std::vector<SynthonSpaceHitSet> results;
+  unsigned int otf;
+  std::vector<std::unique_ptr<ExplicitBitVect>> fragFPs;
+  for (auto &frag : fragSet) {
+    // For the fingerprints, ring info is required.
+    sanitizeMol(*static_cast<RWMol *>(frag.get()), otf,
+                MolOps::SANITIZE_SYMMRINGS);
+    fragFPs.emplace_back(d_fpGen.getFingerprint(*frag));
+  }
+
+  const auto connPatterns = details::getConnectorPatterns(fragSet);
+  boost::dynamic_bitset<> conns(MAX_CONNECTOR_NUM + 1);
+  for (auto &connPattern : connPatterns) {
+    conns |= connPattern;
+  }
+
+  for (const auto &[id, reaction] : getSpace().getReactions()) {
+    // It can't be a hit if the number of fragments is more than the number
+    // of synthon sets because some of the molecule won't be matched in any
+    // of the potential products.  It can be less, in which case the unused
+    // synthon set will be used completely, possibly resulting in a large
+    // number of hits.
+    if (fragSet.size() > reaction->getSynthons().size()) {
+      continue;
+    }
+    // Need to try all combinations of synthon orders.
+    auto synthonOrders =
+        details::permMFromN(fragSet.size(), reaction->getSynthons().size());
+    for (const auto &so : synthonOrders) {
+      // Get all the possible permutations of connector numbers compatible with
+      // the number of synthon sets in this reaction.  So if the
+      // fragmented molecule is C[1*].N[2*] and there are 3 synthon sets
+      // we also try C[2*].N[1*], C[2*].N[3*] and C[3*].N[2*] because
+      // that might be how they're labelled in the reaction database.
+      auto connCombs = details::getConnectorPermutations(
+          fragSet, conns, reaction->getConnectors());
+
+      for (auto &connComb : connCombs) {
+        // All the fragment connectors must match something in the corresponding
+        // synthon.
+        auto connCombConnPatterns = details::getConnectorPatterns(connComb);
+        bool skip = false;
+        for (size_t i = 0; i < connCombConnPatterns.size(); ++i) {
+          if ((connCombConnPatterns[i] & connPatterns[i]).count() <
+              connPatterns[i].count()) {
+            skip = true;
+            break;
+          }
+        }
+        if (skip) {
+          continue;
+        }
+        // It appears that for Morgan fingerprints, the isotope numbers are
```

Comment:
I think that the morgan fingeprints should make use of the isotope values, but something odd is going on. This isn't your problem

Reply:
Oh.  I just performed the experiment, thought "that's handy" and pressed on!

Reply:
Maybe it's ignoring them on dummy atoms?